geode-wan/src/main/java/org/apache/geode/cache/client/internal/locator/wan/LocatorHelper.java

```diff
@@ -46,13 +47,42 @@ public static boolean addLocator(int distributedSystemId, DistributionLocatorId
     Set<DistributionLocatorId> existingValue =
         allLocatorsInfo.putIfAbsent(distributedSystemId, locatorsSet);
     if (existingValue != null) {
-      if (!existingValue.contains(locator)) {
+      if (!locator.getMemberName().equals(DistributionConfig.DEFAULT_NAME)) {
+        DistributionLocatorId existingLocator =
+            getLocatorWithSameMemberName(existingValue, locator);
+
+        if (existingLocator != null) {
+          // if locator with same name exist, check did all parameters are same
+          if (!locator.detailCompare(existingLocator)) {
+            // some parameters had changed for existing locator
+            // replace it
+            existingValue.remove(existingLocator);
+            ConcurrentHashMap<Integer, Set<String>> allServerLocatorsInfo =
+                (ConcurrentHashMap<Integer, Set<String>>) locatorListener
+                    .getAllServerLocatorsInfo();
+            Set<String> alllocators = allServerLocatorsInfo.get(distributedSystemId);
+            alllocators.remove(existingLocator.toString());
+            existingValue.add(locator);
+            addServerLocator(distributedSystemId, locatorListener, locator);
+            locatorListener.locatorJoined(distributedSystemId, locator, sourceLocator);
+            return true;
+          }
+          return false;
+        }
         existingValue.add(locator);
```

Comment:
This code looks wrong to me.  The old code would not invoke addLocator if "existingValue.contains(locator)" returned false.  You still need this check if the check "if (existingLocator != null)" returns false.

if (existingLocator != null) {
  your new code
} else if (!existingValue.contains(locator)) {
  existingValue.add(locator);
  addServerLocator(...);
  locatorListener.locatorJoined(...);
}

The way you've got it coded a new locator with the same host:port but a different memberName will be added as a new locator even if existingValue contains its ID (because equals ignores the member name).  This will cause new threads to be spawned to communicate with that locator, whereas the old code wouldn't do that.

Reply:
This is something that Lynn's test does.  It appends the PID of a locator to its name, so when her test restarts a locator it has the same host:port but a different name.  I didn't think that was bad on first review but it probably accounts for the high number of threads created and subsequent OutOfMemoryError when new threads couldn't be created.

Reply:
Main logic is that member name is main parameter, and if it does not exist, then we apply existing logic.
This is the way we can determine that existing locator has change host[port].

Reply:
Yes, but this has changed the behavior for when the host:port has _not_ changed but the member name _has_ changed.  I think that's what's breaking Lynn's test.

Reply:
The new changes still seem to have the same problem.  If I add a new locator with the same host:port but a **different** member _name_ it will still spawn a thread to invoke locatorJoined in the _else_ clause after "} else if (!existingValue.contains(locator)) {".
Your PR's intent is to scrub an old ID for a new locator that has the **same** member _name_ regardless of its host:port.  If that check fails, meaning there is no ID with the same member _name_, we should then check to see if there is already an ID with the host:port.  If a match is found we should not invoke locatorJoined on the listener. 
 Every invocation of that method spawns a thread and that's causing problems.
Adding some unit tests for these required behaviors would help ensure that things work correctly.

Reply:
I think that test case is not realistic. Is it realistic that locator at restart will change its name. But if this is the case, and also if we are now storing member name, then in that case we should store new member name. I can change this, so test case passes, but this will be inconsistent behavior.

Reply:
hmm, I see what you mean Mario.  The membername field is new, but to be consistent we should invoke the listener in case it cares about that new field.  Hold off on making that change while I look into Lynn's failures further.  Thanks for your patience.